cirq/circuits/circuit_gate.py

```diff
@@ -0,0 +1,384 @@
+            args.append(f'params={self.param_values}')
+        if self.repetitions != 1:
+            args.append(f'loops={self.repetitions}')
+        return f'{base_str}({", ".join(args)})'
+
+    def base_operation(self) -> 'CircuitOperation':
+        """Returns a copy of this operation with only the wrapped gate.
+
+        Key mappings, parameter values, and repetitions are not copied.
+        """
+        return CircuitOperation(self._gate, self._qubits)
+
+    def copy(self) -> 'CircuitOperation':
+        """Returns a copy of this operation."""
+        new_op = self.base_operation()
+        new_op._measurement_key_map = self.measurement_key_map.copy()
+        new_op._param_values = self.param_values.copy()
+        new_op._repetitions = self.repetitions
+        return new_op
+
+    def with_gate(self, new_gate: 'cirq.Gate') -> 'CircuitOperation':
+        """Returns a copy of this operation with the provided gate.
+
+        Key mappings, parameter values, and repetitions are preserved.
+        """
+        if not isinstance(new_gate, CircuitGate):
+            raise TypeError('CircuitOperations may only contain CircuitGates.')
+        if protocols.is_measurement(new_gate) and self._repetitions != 1:
+            raise NotImplementedError(
+                'Measurements cannot be added to looped circuits.')
+        new_op = self.copy()
+        new_op._gate = new_gate
+        return new_op
+
+    def with_measurement_key_mapping(self, key_map: Dict[str, str]
+                                    ) -> 'CircuitOperation':
+        """Returns a copy of this operation with an updated key mapping.
+
+        Parameter values and repetitions are preserved. The provided key_map
+        is composed with the existing map, so calling this function with
+        key_map = {'b': 'c'} on a CircuitOperation with map {'a': 'b'} will
+        result in a final map of {'a': 'c'}.
+        """
+        new_op = self.copy()
+        new_op._measurement_key_map = {
+            k: (key_map[v] if v in key_map else v)
+            for k, v in self.measurement_key_map.items()
+        }
+        new_op._measurement_key_map.update({
+            key: val
+            for key, val in key_map.items()
+            if key not in self.measurement_key_map.values()
+        })
+        return new_op
+
+    def _with_measurement_key_mapping_(self, key_map: Dict[str, str]
+                                      ) -> 'CircuitOperation':
+        return self.with_measurement_key_mapping(key_map)
+
+    def with_params(self, param_values: Dict[sympy.Symbol, Any]
+                   ) -> 'CircuitOperation':
+        """Returns a copy of this operation with an updated key mapping.
+
+        Key mappings and repetitions are preserved. The provided param_values
+        are composed with the existing values, so calling this function with
+        param_values = {'b': 1} on a CircuitOperation with map {'a': 'b'} will
+        result in a final map of {'a': 1}.
```

Comment:
This is somewhat confusing to me. I wonder if it would be more clear to maintain two separate maps, a "symbol map" that is a bi-directional mapping between "inner" and "outer" symbols (like the measurement key map) and a separate dict of "resolved" outer parameters.

Maybe it would help to actually implement the parameter resolution protocol methods `_resolve_parameters_` and `_parameter_names_` because it's not clear how `with_params` fits in with parameter resolution.

Reply:
This map assigns values to "inner" parameters, not "outer" ones. The example makes more sense in a nested context, e.g.:
```
Circuit(
  CircuitGate(
    X(k) ** b,
    CircuitGate(
      X(k) ** a,
    ).on(k).with_params('a': 'b')
  ).on(k).with_params('b': 1)
)
```

Its purpose is to defer applying parameter values until the `CircuitOperation` is decomposed, similar to the measurement key mapping. This helps with serialization and will open up some dynamic behavior possibilities once classical control is supported.

Reply:
The parameter map conflates two things currently:
- "rebinding" symbols so that you can embed the same circuit multiple times and vary the parameters on each one independently (e.g. embed `CircuitGate(X(k) ** a)` twice but rebind the exponent to separate symbols `a0` and `a1` in the "outer" circuit).
- assigning values to parameters (actual resolution is deferred until later so that we can reuse embedded circuit instances, as you said).

I'm wondering if we should explicitly separate those two in the code, instead of 'collapsing' them into one map as in the current implementation. It would definitely help to implement the parameter resolution special methods to make all this more clear.

One thing that is a bit odd in the current implementation is that it looks like one could "unresolve" parameters. For example:
```python
g = CircuitGate(X(k) ** a).on(k).with_params({'a': 1})
cirq.is_parameterized(g)  # => False (need to implement this)
cirq.parameter_names(g)  # => set()  (

g2 = g.with_params({'a': 'b'})
cirq.is_parameterized(g2)  # => True
```
That behavior feels a bit odd to me.

Reply:
I'm currently adding support for `ParamResolver` composition in #3546 and #3538. Following those changes, `CircuitOperation` will use a `ParamResolver` with single-step resolution for its param value mapping.